File: cam_mon_java/src/main/java/com/marble/cammon/TestServoApi.java
package com.marble.cammon;

public class TestServoApi {
    public static void main(String[] args) throws Exception {
        Protocol.ServoPacket s = new Protocol.ServoPacket();
        s.seq = 1;
        s.deviceType = 1;
        s.packetType = 2;
        s.deviceIp = 0;
        s.reservedConn = 0;
        s.control = 0x11;
        s.reservedCtrl = 0;
        s.azimuth = 123.45f;
        s.elevation = 10.0f;
        s.azSpeed = 1.5f;
        s.elSpeed = 0.5f;
        s.targetDistance = 100;
        s.reserved = new byte[33];
        s.reservedHorz = 0;
        s.reservedVert = 0;
        s.timestamp = 0;
        s.reservedState = 0;
        s.backup = 0;
        byte[] out = s.serializeServo();
        System.out.println("Serialized servo length=" + out.length);
        Protocol.ServoPacket parsed = Protocol.ServoPacket.parseServo(out);
        if (parsed == null) {
            System.err.println("parse failed");
            System.exit(2);
        }
        if (Math.abs(parsed.azimuth - s.azimuth) > 1e-5) { System.err.println("az mismatch"); System.exit(3); }
        if (Math.abs(parsed.elevation - s.elevation) > 1e-5) { System.err.println("el mismatch"); System.exit(4); }
        if (Math.abs(parsed.azSpeed - s.azSpeed) > 1e-5) { System.err.println("azs mismatch"); System.exit(5); }
        if (Math.abs(parsed.elSpeed - s.elSpeed) > 1e-5) { System.err.println("els mismatch"); System.exit(6); }
        if (parsed.targetDistance != s.targetDistance) { System.err.println("dist mismatch"); System.exit(7); }
        System.out.println("Servo API test passed");
    }
}

File: cam_mon_cpp/src/test_servo_api.cpp
#include "protocol.h"
#include <vector>
#include <iostream>
#include <cmath>

using namespace cammon;

int main() {
    float az = 123.45f;
    float el = 10.0f;
    float azs = 1.5f;
    float els = 0.5f;
    uint16_t dist = 100;
    auto out = build_servo_packet(az, el, azs, els, dist, /*seq=*/1, /*control=*/0x11);
    std::cout << "Serialized servo length=" << out.size() << "\n";
    auto sp = ServoPacket::deserialize_servo(out);
    if (!sp) { std::cerr << "deserialize failed\n"; return 2; }
    // compare values with small epsilon
    if (std::fabs(sp->azimuth - az) > 1e-5) { std::cerr << "az mismatch: " << sp->azimuth << " vs " << az << "\n"; return 3; }
    if (std::fabs(sp->elevation - el) > 1e-5) { std::cerr << "el mismatch\n"; return 4; }
    if (std::fabs(sp->az_speed - azs) > 1e-5) { std::cerr << "azs mismatch\n"; return 5; }
    if (std::fabs(sp->el_speed - els) > 1e-5) { std::cerr << "els mismatch\n"; return 6; }
    if (sp->target_distance != dist) { std::cerr << "dist mismatch\n"; return 7; }
    std::cout << "Servo API test passed\n";
    return 0;
}
